src/controllers/VueController.php

```diff
@@ -191,20 +311,20 @@ private function _getStats()
                 'totalOrders' => [
                     'total' => $numCurrentOrders,
                     // this is based on the new previous period data
-                    'percentChange' => round((($numCurrentOrders - $numPreviousOrders) / $numPreviousOrders) * 100, 2),
-                    'revenue' => round((($currentRevenue - $previousRevenue) / $previousRevenue) * 100, 2),
-                    'series' => [32, 40, 43, 45, 300, 56, 60, 80, 90, 105]
+                    'percentChange' => $numPreviousOrders ? round((($numCurrentOrders - $numPreviousOrders) / $numPreviousOrders) * 100, 2) : 0,
+                    'revenue' => $previousRevenue ? round((($currentRevenue - $previousRevenue) / $previousRevenue) * 100, 2) : 0,
+                    'series' => $totalOrdersSet
                 ],
                 // averageOrderValue, averageOrderQuantity
                 'averageValue' => [
-                    'total' => 98.76,
-                    'percentChange' => -3,
-                    'series' => [32, 40, 43, 45, 49, 56, 60, 80, 90, 105]
+                    'total' => $numCurrentOrders ? round($currentRevenue / $numCurrentOrders, 2) : 0,
```

Comment:
When working with currency, we should at least be using integers and not floating point types whereby the integer value would represent the total in cents. This avoids any floating-point errors in the calculations and leaves less chance to be off by a cent.

Using floating point type:

```php
$total = 100000000;
for ($i = 1; $i < 100000000; $i++) {
 $total = $total - 0.99;
}
echo round($total, 2);
// Prints 1000001.12
```

Integer type:

```php
// Total represented in cents
$total = 10000000000;
for ($i = 1; $i < 100000000; $i++) {
  // Subtraction amount in cents
	$total = $total - 99;
}
echo $total/100;
// Prints 1000000.99
```

Even better, using a well-supported package specifically for handling currencies: https://github.com/moneyphp/money

Reply:
Does Craft Commerce already come with a currency package?

Reply:
Yes, they have [`"^3.2.1"`](https://github.com/craftcms/commerce/blob/2451d0c6/composer.json#L25), although I can't see that its actually being used anywhere in Commerce. It was added years ago - https://github.com/craftcms/commerce/commit/795d3126ed6c73e4e23fb7ad6f1b0b5730c39867.

We would do well to require the same version as Commerce does.

Reply:
I don't think that'll work for us. It's called this way: `new Money($dollarAmt, new Currency('USD'))`, where, in this case, the `Currency` class interferes with Commerce's `Currency` class because we don't need 2 classes for currency, and it's better to use the one Commerce uses. The other way of doing it would be `Money::USD($dollarAmt)`, but this isn't a great solution either because the currency we want to use could actually be anything, and in many sites can easily be a mixture of multiple currencies.

Reply:
OK, but this ignores the problem here. The currency handling is insufficient - in its current form it will suffer from floating-point precision errors. See my example above.

RE Craft's Currency class. They have _two_ classes, one is a [model](https://github.com/craftcms/commerce/blob/2451d0c6/src/models/Currency.php), the other is a [helper](https://github.com/craftcms/commerce/blob/2451d0c6/src/helpers/Currency.php). 

We don't need to worry about conflicts with the classes because

1. We won't be passing moneyphp objects to Craft internals, instead we use that in _our_ logic;
2. Where we make use of Craft Commerce's classes named `Currency` we can either alias the Commerce classes or the moneyphp classes. 
  - Please refer to PHP's aliasing documentation: https://www.php.net/manual/en/language.namespaces.importing.php

```php
use craft\commerce\helpers\Currency as CurrencyHelper;
use Money\Money;
use Money\Currency;

...

    // Somewhere in a function
    new Money($dollarAmt, new Currency('USD'))

    protected static function convertCurrency(float $amount, string $currency, bool $convert = true) : string {
        $amt = CurrencyHelper::formatAsCurrency($amount, $currency, $convert);

        if (strpos($amt, '-')) {
            $amt = str_replace('-', '', $amt);
            $amt = '-' . $amt;
        }
        return $amt;
    }
```

I recommend using the moneyphp library than handling currency conversions/maths yourself.